server/app/services/pubsub/nats.rb

```diff
@@ -0,0 +1,49 @@
+require 'nats/io/client'
+require_relative 'base'
+
+module Pubsub
+  class Nats < Base
+
+    CHANNEL = 'master_pubsub'.freeze
+
+    # @param servers [Array<String>]
+    def initialize(servers)
+      @servers = servers
+      @subscriptions = []
+      @nats = NATS::IO::Client.new
+      @nats.on_error do |e|
+        error(e)
+      end
+      @nats.on_reconnect do
+        info "reconnected to server at #{@nats.connected_server}"
+      end
+      @nats.on_close do
+        warning "connection to NATS closed"
+      end
+      async.stream!
+    end
+
+    # @param channel [String]
+    # @param data [Hash]
+    def publish(channel, data)
+      msg = { channel: channel, data: data }
+      @nats.publish(CHANNEL, MessagePack.pack(msg))
```

Comment:
I'm not sure if multiplexing all pubsub channels over a single backend channel is the optimal implementation in general, but I don't know how NATS behaves specifically.

I think the relevant question is whether our pubsub interface needs to preserve message ordering across different pubsub channels, and I don't think that's the case... pubsub messages should be allowed to be re-ordered across different channels, and I think that might also be a performance win.

Reply:
> I'm not sure if multiplexing all pubsub channels over a single backend channel is the optimal implementation in general, but I don't know how NATS behaves specifically.

Single backend channel is much more simpler to implement vs multiple backend channels.. with multiple channels we get channel discovery complexity etc.